docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,342 @@
+template take precedence over the same parameters on the instances.
+
+The main difference with [Option 1] that enables ClusterClass is that the same ManagedCluster template
+resource can be referenced by multiple AzureManagedControlPlanes, so this new `ASOManagedClusterTemplateRef`
+field can be defined on the AzureManagedControlPlaneClassSpec so a set of AKS parameters defined once in the
+template can be applied to all Clusters built from that ClusterClass.
+
+This method makes all ManagedCluster fields available to define in a template which could lead to
+misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a
+template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ
+will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and
+omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known
+fields which could or should not be defined in template types and will otherwise rely on AKS to provide
+reasonable error messages for misconfigurations.
+
+Like [Option 1], this method keeps particular versions of CAPZ decoupled from particular API versions of ASO
+resources (including allowing preview versions) and opens the door for streamlined adoption of existing AKS
+clusters.
+
+#### Option 3: CAPZ resource defines an entire unstructured ASO resource inline
+
+This method is functionally equivalent to [Option 2] except that the template resource is defined inline
+within the AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterTemplate is the ASO ManagedCluster to be used as a template from which new
+	// ManagedClusters will be created.
+	ManagedClusterTemplate map[string]interface{} `json:"managedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
+as an ASO ManagedCluster:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterSpec defines the spec of the ASO ManagedCluster managed by this AzureManagedControlPlane.
+	ManagedClusterSpec v1api20230201.ManagedCluster_Spec `json:"managedClusterTemplate"`
```

Comment:
One potentially significant issue I could see with this approach is that since ASO implements conversion webhooks between each API version of its resources, the most correct thing for CAPZ to do would probably be to also invoke those same webhooks as we update the ASO API version here which I think we could only practically do unless we also cut a new API version for AzureManagedControlPlane each time. One API version per AKS API update seems like it would get pretty unwieldy.

Reply:
> One API version per AKS API update seems like it would get pretty unwieldy.

Yes, we should try to avoid this

Reply:
@matthchr How critical do you think this conversion webhook concern is?

Reply:
Oh sorry somehow I missed this question and this is a very important question.

There _is_ custom conversion logic for managed cluster versions especially between preview API versions as there can be changes there and we do write custom conversion to tolerate it. If you just roll the version without calling out conversion there's some significant risk there. GA APIs are safer but not immune (see for example the `Basic` vs `Base`, `Standard` vs `Paid`) change that happened a bit ago. ASO does specific conversion to make that work across API versions in a way users don't need to care about.

Again you might be able to call our webhook... or else we could somehow keep you appraised of those customizations and possibly make them reusable for you somehow...

Reply:
And all of those workarounds would still require a separate CAPZ API version anytime we bump the ASO API version embedded in here, right?

Reply:
I don't fully understand/know what the CAPZ expectation for when to produce a new API version is. Is it only when there are breaking changes and/or conversion webhooks are needed? Then maybe you could get away with changes that are pure-addition without needing a new CAPZ version.

If there are breaking changes you definitely need a new version + conversion webhook to make it work.

The fact that you don't have control over when there are breaking changes is one of the reasons why ASO puts the API version in the CRD version, because that way we can use standard conversion webhooks to resolve breaking changes ideally without needing user input. The problem is that most k8s versioning doesn't work that way and ASO is "weird" for doing it - if/when you get to `v1` what do you do when AKS (and thus ASO) ships a breaking change? Do you go to `v2`?

Reply:
> Is there a good way to figure out whether a new ASO API version includes only additive changes? Or would we have to inspect the old and new types and conversion webhooks to know for sure?

There are ways but there are no good ways that I can think of. I think you'd have to inspect the webhook impl and/or the shape of the N-1 and N versions.